File: shared/src/commonMain/kotlin/org/nihongo/mochi/domain/settings/SettingsRepository.kt
package org.nihongo.mochi.domain.settings

import com.russhwolf.settings.Settings
import org.nihongo.mochi.domain.services.VoiceGender
import org.nihongo.mochi.settings.*

class SettingsRepository(private val settings: Settings) {

    fun isFirstRun(): Boolean = settings.getBoolean(IS_FIRST_RUN_PREF_KEY, true)
    fun setFirstRunCompleted() = settings.putBoolean(IS_FIRST_RUN_PREF_KEY, false)

    fun hasSeenRecapTutorial(): Boolean = settings.getBoolean("has_seen_recap_tutorial", false)
    fun setRecapTutorialSeen() = settings.putBoolean("has_seen_recap_tutorial", true)

    fun hasSeenKanaTutorial(): Boolean = settings.getBoolean("has_seen_kana_tutorial", false)
    fun setKanaTutorialSeen() = settings.putBoolean("has_seen_kana_tutorial", true)
    
    fun hasSeenKanjiDetailTutorial(): Boolean = settings.getBoolean("has_seen_kanji_detail_tutorial", false)
    fun setKanjiDetailTutorialSeen() = settings.putBoolean("has_seen_kanji_detail_tutorial", true)

    fun hasSeenHomeTutorial(): Boolean = settings.getBoolean("has_seen_home_tutorial", false)
    fun setHomeTutorialSeen() = settings.putBoolean("has_seen_home_tutorial", true)

    fun getPronunciation(): String = settings.getString(PRONUNCIATION_PREF_KEY, "Hiragana")
    fun setPronunciation(value: String) = settings.putString(PRONUNCIATION_PREF_KEY, value)

    fun getAnimationSpeed(): Float = settings.getFloat(ANIMATION_SPEED_PREF_KEY, 1.0f)
    fun setAnimationSpeed(value: Float) = settings.putFloat(ANIMATION_SPEED_PREF_KEY, value)

    fun getTextSize(): Float = settings.getFloat(TEXT_SIZE_PREF_KEY, 1.0f)
    fun setTextSize(value: Float) = settings.putFloat(TEXT_SIZE_PREF_KEY, value)

    fun shouldAddWrongAnswers(): Boolean = settings.getBoolean(ADD_WRONG_ANSWERS_PREF_KEY, true)
    fun setAddWrongAnswers(value: Boolean) = settings.putBoolean(ADD_WRONG_ANSWERS_PREF_KEY, value)

    fun shouldRemoveGoodAnswers(): Boolean = settings.getBoolean(REMOVE_GOOD_ANSWERS_PREF_KEY, true)
    fun setRemoveGoodAnswers(value: Boolean) = settings.putBoolean(REMOVE_GOOD_ANSWERS_PREF_KEY, value)
    
    fun getTheme(): String = settings.getString(THEME_PREF_KEY, "light")
    fun setTheme(value: String) = settings.putString(THEME_PREF_KEY, value)
    
    fun getAppLocale(): String = settings.getString(APP_LOCALE_PREF_KEY, "en_GB")
    fun setAppLocale(value: String) = settings.putString(APP_LOCALE_PREF_KEY, value)

    fun getMode(): String = settings.getString(MODE_PREF_KEY, "JLPT")
    fun setMode(value: String) = settings.putString(MODE_PREF_KEY, value)

    fun getSelectedLevel(): String = settings.getString(SELECTED_LEVEL_PREF_KEY, "")
    fun setSelectedLevel(value: String) = settings.putString(SELECTED_LEVEL_PREF_KEY, value)

    // Game States Persistance
    fun saveGameState(gameKey: String, json: String) = settings.putString(gameKey, json)
    fun getGameState(gameKey: String): String? = settings.getStringOrNull(gameKey)
    fun clearGameState(gameKey: String) = settings.remove(gameKey)

    // TTS Settings
    fun getTtsGender(locale: String): VoiceGender {
        if (locale.startsWith("ar"))
            return VoiceGender.MALE
        val defaultGender = VoiceGender.FEMALE.name
        val genderName = settings.getString(TTS_GENDER_PREF_KEY, defaultGender)
        return try {
            VoiceGender.valueOf(genderName)
        } catch (e: Exception) {
            VoiceGender.valueOf(defaultGender)
        }
    }
    
    fun setTtsGender(gender: VoiceGender) = settings.putString(TTS_GENDER_PREF_KEY, gender.name)

    fun getTtsRate(): Float = settings.getFloat(TTS_RATE_PREF_KEY, 1.0f)
    fun setTtsRate(rate: Float) = settings.putFloat(TTS_RATE_PREF_KEY, rate)

    fun getTtsVoiceId(): String? = settings.getStringOrNull(TTS_VOICE_ID_PREF_KEY)
    fun setTtsVoiceId(voiceId: String?) = settings.putString(TTS_VOICE_ID_PREF_KEY, voiceId ?: "")
}

File: flutter_app/lib/repositories/settings_repository.dart
import 'package:shared_preferences/shared_preferences.dart';

class SettingsRepository {
  final SharedPreferences _prefs;

  SettingsRepository(this._prefs);

  // Clés identiques au projet Kotlin pour référence
  static const String _isFirstRunKey = "is_first_run";
  static const String _pronunciationKey = "pronunciation";
  static const String _animationSpeedKey = "animation_speed";
  static const String _themeKey = "theme_pref";
  static const String _appLocaleKey = "app_locale";
  static const String _modeKey = "mode_pref";
  static const String _audioVolumeKey = "audio_volume";
  static const String _ttsRateKey = "tts_rate";
  static const String _addWrongAnswersKey = "add_wrong_answers";
  static const String _removeGoodAnswersKey = "remove_good_answers";
  static const String _selectedLevelKey = "selected_level";
  static const String _ttsVoiceIdKey = "tts_voice_id";

  bool isFirstRun() => _prefs.getBool(_isFirstRunKey) ?? true;
  Future<void> setFirstRunCompleted() => _prefs.setBool(_isFirstRunKey, false);

  String getPronunciation() => _prefs.getString(_pronunciationKey) ?? "Hiragana";
  Future<void> setPronunciation(String value) => _prefs.setString(_pronunciationKey, value);

  double getAnimationSpeed() => _prefs.getDouble(_animationSpeedKey) ?? 1.0;
  Future<void> setAnimationSpeed(double value) => _prefs.setDouble(_animationSpeedKey, value);

  String getTheme() => _prefs.getString(_themeKey) ?? "light";
  Future<void> setTheme(String value) => _prefs.setString(_themeKey, value);

  String getAppLocale() => _prefs.getString(_appLocaleKey) ?? "en_GB";
  Future<void> setAppLocale(String value) => _prefs.setString(_appLocaleKey, value);

  String getMode() => _prefs.getString(_modeKey) ?? "JLPT";
  Future<void> setMode(String value) => _prefs.setString(_modeKey, value);

  double getAudioVolume() => _prefs.getDouble(_audioVolumeKey) ?? 1.0;
  Future<void> setAudioVolume(double value) => _prefs.setDouble(_audioVolumeKey, value);

  double getTtsRate() => _prefs.getDouble(_ttsRateKey) ?? 1.0;
  Future<void> setTtsRate(double value) => _prefs.setDouble(_ttsRateKey, value);

  bool shouldAddWrongAnswers() => _prefs.getBool(_addWrongAnswersKey) ?? true;
  Future<void> setAddWrongAnswers(bool value) => _prefs.setBool(_addWrongAnswersKey, value);

  bool shouldRemoveGoodAnswers() => _prefs.getBool(_removeGoodAnswersKey) ?? true;
  Future<void> setRemoveGoodAnswers(bool value) => _prefs.setBool(_removeGoodAnswersKey, value);

  String getSelectedLevel() => _prefs.getString(_selectedLevelKey) ?? "";
  Future<void> setSelectedLevel(String value) => _prefs.setString(_selectedLevelKey, value);

  String? getTtsVoiceId() => _prefs.getString(_ttsVoiceIdKey);
  Future<void> setTtsVoiceId(String? value) {
    if (value == null || value.isEmpty) {
      return _prefs.remove(_ttsVoiceIdKey);
    }
    return _prefs.setString(_ttsVoiceIdKey, value);
  }
}

